In [1]:
lang = {
            'aa' : 'afar',
            'ab' : 'abjasio',  # (o abjasiano)
            'ae' : 'avéstico',
            'af' : 'afrikáans',
            'ak' : 'akano',
            'am' : 'amhárico',
            'an' : 'aragonés',
            'ar' : 'العربية',
            'as' : 'asamés',
            'av' : 'avar',  # (o ávaro)
            'ay' : 'aimara',
            'az' : 'azerí',
            'ba' : 'baskir',
            'be' : 'беларуская мова',
            'bg' : 'български език',
            'bh' : 'bhoyapurí',
            'bi' : 'bislama',
            'bm' : 'bamanankan',
            'bn' : 'বাংলা',
            'bo' : 'tibetano',
            'br' : 'bretón',
            'bs' : 'bosanski',
            'ca' : 'català',
            'ce' : 'checheno',
            'ch' : 'chamorro',
            'co' : 'corso',
            'cr' : 'cree',
            'cs' : 'český',
            'cu' : 'eslavo',  # eclesiástico antiguo
            'cv' : 'chuvasio',
            'cy' : 'cymraeg',
            'da' : 'dansk',
            'de' : 'deutsch',
            'dv' : 'maldivo',  # (o dhivehi)
            'dz' : 'dzongkha',
            'ee' : 'ewé',
            'el' : 'ελληνικά',  # (moderno)
            'en' : 'english',
            'eo' : 'esperanto',
            'es' : 'español',  # (o castellano)
            'et' : 'eesti',
            'eu' : 'euskera',
            'fa' : 'فارسی',
            'ff' : 'fula',
            'fi' : 'suomi',  # (o finlandés)
            'fj' : 'fiyiano',  # (o fiyi)
            'fo' : 'feroés',
            'fr' : 'français',
            'fy' : 'frisón',  # (o frisio)
            'ga' : 'gaeilge',  # (o gaélico)
            'gd' : 'gaélico',  # escocés
            'gl' : 'galego',
            'gn' : 'guaraní',
            'gu' : 'guyaratí',  # (o gujaratí)
            'gv' : 'manés',  # (gaélico manés o de Isla de Man)
            'ha' : 'hausa',
            'he' : 'עִבְרִית',
            'hi' : 'हिन्दी',  # (o hindú)
            'ho' : 'hiri',  # motu
            'hr' : 'hrvatski',
            'ht' : 'haitiano',
            'hu' : 'magyar',
            'hy' : 'armenio',
            'hz' : 'herero',
            'ia' : 'interlingua',
            'id' : 'bahasa indonesia',
            'ie' : 'occidental',
            'ig' : 'igbo',
            'ii' : 'yi de Sichuán',
            'ik' : 'iñupiaq',
            'io' : 'ido',
            'is' : 'íslenska',
            'it' : 'italiano',
            'iu' : 'inuktitut',  # (o inuit)
            'ja' : '日本語',
            'jv' : 'javanés',
            'ka' : 'ქართული',
            'kg' : 'kongo',  # (o kikongo)
            'ki' : 'kikuyu',
            'kj' : 'kuanyama',
            'kk' : 'қазақ',  # (o kazajio)
            'kl' : 'groenlandés',  # (o kalaallisut)
            'km' : 'camboyano',  # o jemer)
            'kn' : 'Kannada',
            'ko' : '한국어/조선말',
            'kr' : 'kanuri',
            'ks' : 'cachemiro',  # (o cachemir)
            'ku' : 'kurdo',
            'kv' : 'komi',
            'kw' : 'córnico',
            'ky' : 'kirghiz',
            'la' : 'latín',
            'lb' : 'luxemburgués',
            'lg' : 'luganda',
            'li' : 'limburgués',
            'ln' : 'lingala',
            'lo' : 'lao',
            'lt' : 'lietuvi\x9akai',
            'lu' : 'luba-katanga',  # (o chiluba)
            'lv' : 'latviešu',
            'mg' : 'malgache',  # (o malagasy)
            'mh' : 'marshalés',
            'mi' : 'maorí',
            'mk' : 'macedonio',
            'ml' : 'malayalam',
            'mn' : 'mongol',
            'mr' : 'maratí',
            'ms' : 'bahasa melayu',
            'mt' : 'malti',
            'my' : 'birmano',
            'na' : 'nauruano',
            'nb' : 'bokmål',  # bokmål
            'nd' : 'ndebele del norte',
            'ne' : 'nepalí',
            'ng' : 'ndonga',
            'nl' : 'nederlands',  # (u holandés)
            'nn' : 'nynorsk',
            'no' : 'norsk',
            'nr' : 'ndebele del sur',
            'nv' : 'navajo',
            'ny' : 'chichewa',
            'oc' : 'occitano',
            'oj' : 'ojibwa',
            'om' : 'oromo',
            'or' : 'oriya',
            'os' : 'osético',  # (u osetio, u oseta)
            'pa' : 'ਪੰਜਾਬੀ', # (o penyabi)
            'pi' : 'pali',
            'pl' : 'polski',
            'ps' : 'پښتو',  # (o pastún, o pashto)
            'pt' : 'português',
            'qc' : 'quechua',
            'rm' : 'romanche',
            'rn' : 'kirundi',
            'ro' : 'română',
            'ru' : 'pусский',
            'rw' : 'kinyarwanda',  # (o kiñaruanda)
            'sa' : 'sánscrito',
            'sc' : 'sardo',
            'sd' : 'sindhi',
            'se' : 'sami septentrional',
            'sg' : 'sango',
            'si' : 'cingalés',
            'sk' : 'slovenčina',
            'sl' : 'slovenščina',
            'sm' : 'samoano',
            'sn' : 'shona',
            'so' : 'somalí',
            'sq' : 'shqip',
            'sr' : 'srpski',
            'ss' : 'suazi',  # (o swati, o siSwati)
            'st' : 'sesotho',
            'su' : 'sundanés',  # (o sondanés)
            'sv' : 'svenska',
            'sw' : 'kiswahili',
            'ta' : 'தமிழ்',
            'te' : 'తెలుగు',
            'tg' : 'tayiko',
            'th' : 'ภาษาไทย',
            'ti' : 'tigriña',
            'tk' : 'turcomano',
            'tl' : 'tagalo',
            'tn' : 'setsuana',
            'to' : 'tongano',
            'tr' : 'türkçe',
            'ts' : 'tsonga',
            'tt' : 'tártaro',
            'tw' : 'twi',
            'ty' : 'tahitiano',
            'ug' : 'uigur',
            'uk' : 'український',
            'ur' : 'اردو',
            'uz' : 'ozbek',
            've' : 'venda',
            'vi' : 'tiếng việt',
            'vo' : 'volapük',
            'wa' : 'valón',
            'wo' : 'wolof',
            'xh' : 'xhosa',
            'yi' : 'yídish',  # (o yidis, o yiddish)
            'yo' : 'yoruba',
            'za' : 'chuan',  # (o chuang, o zhuang)
            'zh' : '普通话',
            'zu' : 'isizulu',
            'cn' : '广州话 / 廣州話',
            'Sin Idioma': 'no language'
            }

# ETL (Extract, Transform, Load)

In [2]:
# Importar modulos necesarios
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

# Importar función 'literal_eval' desde módulo ast
from ast import literal_eval

## Funciones Personalizadas (custom functions)

In [3]:
# Función para verificar si existen valores no deseados en campos numéricos
def not_num(array, remp=0):
    cart = []
    for n in array:
        try:
            float(n)
        except:
            cart.append(n)
    return cart

def flatten_column(df, column, key = 'name'):
    
    # Convertir todos los valores faltantes en cadenas de listas vacías
    df[column] = df[column].fillna('[]')

    # Aplicar función literal_eval para convertir los objetos con datos tipo string en objectos tipo dict
    df[column] = df[column].apply(literal_eval)
    
    for e in list(df[column]):
        if len(e) > 0:
            tipo = type(e)
            break
    if tipo == dict:
        # Convertir las listas de diccionarios en listas de string
        df[column] = df[column].apply(lambda x : x[key].lower() if isinstance(x, dict) else 'no collection')

    elif tipo == list:
        # Convertir las listas de diccionarios en listas de string
        df[column] = df[column].apply(lambda x: [i[key].lower() for i in x] if isinstance(x, list) else '')

    else:
        print('Tipo de dato no soportado')
            
# Extraer Nombre del Director si no existe retorna nan
def get_director(x):
    for crew_member in x:
        if crew_member['job'] == 'Director':
            return crew_member['name']
    return np.nan

def inc_iso639(x):
    if len(x)>2:
        for k, v in lang.items():
            if v == x.lower():
                return k
    else:
        for k, v in lang.items():

            if k == x.lower():
                return v

In [4]:
# Carga de datasets
movies = pd.read_csv('datasets/movies_dataset.csv')
credits = pd.read_csv('datasets/credits.csv')

/tmp/ipykernel_9704/3097283174.py:2: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  movies = pd.read_csv('datasets/movies_dataset.csv')


## 1. Visualización Inicial

In [5]:
print("Forma de dataframe movies: ",movies.shape)
print("Forma de dataframe credits: ", credits.shape)

Forma de dataframe movies:  (45466, 24)
Forma de dataframe credits:  (45476, 3)


In [6]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  45466 non-null  object 
 1   belongs_to_collection  4494 non-null   object 
 2   budget                 45466 non-null  object 
 3   genres                 45466 non-null  object 
 4   homepage               7782 non-null   object 
 5   id                     45466 non-null  object 
 6   imdb_id                45449 non-null  object 
 7   original_language      45455 non-null  object 
 8   original_title         45466 non-null  object 
 9   overview               44512 non-null  object 
 10  popularity             45461 non-null  object 
 11  poster_path            45080 non-null  object 
 12  production_companies   45463 non-null  object 
 13  production_countries   45463 non-null  object 
 14  release_date           45379 non-null  object 
 15  re

In [7]:
# Chequeando valores nulos por cada columna en porcentaje
movies.isnull().sum()*100/movies.shape[0]

adult                     0.000000
belongs_to_collection    90.115691
budget                    0.000000
genres                    0.000000
homepage                 82.883913
id                        0.000000
imdb_id                   0.037391
original_language         0.024194
original_title            0.000000
overview                  2.098271
popularity                0.010997
poster_path               0.848986
production_companies      0.006598
production_countries      0.006598
release_date              0.191352
revenue                   0.013197
runtime                   0.578454
spoken_languages          0.013197
status                    0.191352
tagline                  55.104914
title                     0.013197
video                     0.013197
vote_average              0.013197
vote_count                0.013197
dtype: float64

In [8]:
credits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45476 entries, 0 to 45475
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   cast    45476 non-null  object
 1   crew    45476 non-null  object
 2   id      45476 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 1.0+ MB


In [9]:
# Chequeando valores nulos por cada columna en porcentaje
credits.isnull().sum()*100/credits.shape[0]

cast    0.0
crew    0.0
id      0.0
dtype: float64

### Observaciones Vizualización Inicial:

* DataFrame: `movies`
    * Cantidad de Filas: 45466
    * Cantidad de Columnas: 24
    * Tipos de datos: float64(4), object(20)
    * Uso en memoria: 8.3+ MB
    * Detalles a tomar en cuenta
        * Columna `belongs_to_collection`presenta 40972 datos faltantes representa el 90.11%
        * Columna `homepage`presenta 37684 datos faltantes representa el 82.88%
        * Columna `tagline`presenta 25054 datos faltantes representa el 55.11%
        * Las columnas `belongs_to_collection`, `genres`, `spoken_languages` son columnas con datos anidados.
        * La columna `id` se presenta con tipo de dato object


* DataFrame: `credits`
    * Cantidad de Filas: 45476
    * Cantidad de Columnas: 3
    * Tipos de datos: int64(1), object(2)
    * Uso en memoria: 1.0+ MB
    * Detalles a tomar en cuenta
        * No existen datos faltantes
        * Las columnas `cast`, `crew` son columnas con datos anidados.

## 2. Limpieza de Datos

### 2.1 Eliminar filas duplicadas en df movies

In [10]:
print(f'Tamaño del set antes de eliminar las filas repetidas: {movies.shape}')
movies.drop_duplicates(inplace=True)
print(f'Tamaño del set después de eliminar las filas repetidas: {movies.shape}')

Tamaño del set antes de eliminar las filas repetidas: (45466, 24)
Tamaño del set después de eliminar las filas repetidas: (45453, 24)


### 2.2 Eliminar filas duplicadas en df credits

In [11]:
print(f'Tamaño del set antes de eliminar las filas repetidas: {credits.shape}')
credits.drop_duplicates(inplace=True)
print(f'Tamaño del set después de eliminar las filas repetidas: {credits.shape}')

Tamaño del set antes de eliminar las filas repetidas: (45476, 3)
Tamaño del set después de eliminar las filas repetidas: (45439, 3)


### 2.3 Verificar valores no numéricos en columna `id`de df movies

In [12]:
# Verificar valores no numéricos en columna "id"
not_num(movies.id.values)

['1997-08-20', '2012-09-29', '2014-01-01']

#### 2.3.1 Verificar registros con valores anómalos en la columna `id` y si sus datos pueden ser útiles o rescatables

In [13]:
movies[movies['id'] == '1997-08-20']

adult belongs_to_collection  \
19730   - Written by Ørnås              0.065736   

                                 budget  \
19730  /ff9qCepilowshEtG2GYWwzt2bs4.jpg   

                                                  genres  \
19730  [{'name': 'Carousel Productions', 'id': 11176}...   

                                                homepage          id imdb_id  \
19730  [{'iso_3166_1': 'CA', 'name': 'Canada'}, {'iso...  1997-08-20       0   

      original_language                            original_title  overview  \
19730             104.0  [{'iso_639_1': 'en', 'name': 'English'}]  Released   

       ... release_date revenue runtime spoken_languages status  tagline  \
19730  ...            1     NaN     NaN              NaN    NaN      NaN   

       title video vote_average vote_count  
19730    NaN   NaN          NaN        NaN  

[1 rows x 24 columns]

In [14]:
movies[movies['id'] == '2012-09-29']

adult  \
29503   Rune Balot goes to a casino connected to the ...   

      belongs_to_collection                            budget  \
29503              1.931659  /zV8bHuSL6WXoD6FWogP9j4x80bL.jpg   

                                                  genres  \
29503  [{'name': 'Aniplex', 'id': 2883}, {'name': 'Go...   

                                                homepage          id imdb_id  \
29503  [{'iso_3166_1': 'US', 'name': 'United States o...  2012-09-29       0   

      original_language                        original_title  overview  ...  \
29503              68.0  [{'iso_639_1': 'ja', 'name': '日本語'}]  Released  ...   

      release_date revenue runtime spoken_languages status  tagline  title  \
29503           12     NaN     NaN              NaN    NaN      NaN    NaN   

      video vote_average vote_count  
29503   NaN          NaN        NaN  

[1 rows x 24 columns]

In [15]:
movies[movies['id'] == '2014-01-01']

adult  \
35587   Avalanche Sharks tells the story of a bikini ...   

      belongs_to_collection                            budget  \
35587              2.185485  /zaSf5OG7V8X8gqFvly88zDdRm46.jpg   

                                                  genres  \
35587  [{'name': 'Odyssey Media', 'id': 17161}, {'nam...   

                                       homepage          id imdb_id  \
35587  [{'iso_3166_1': 'CA', 'name': 'Canada'}]  2014-01-01       0   

      original_language                            original_title  overview  \
35587              82.0  [{'iso_639_1': 'en', 'name': 'English'}]  Released   

       ... release_date revenue runtime spoken_languages status  tagline  \
35587  ...           22     NaN     NaN              NaN    NaN      NaN   

       title video vote_average vote_count  
35587    NaN   NaN          NaN        NaN  

[1 rows x 24 columns]

#### 2.3.2 Eliminar columnas con valores anómalos en la columna `id` ya que no pueden ser rescatables

In [16]:
# Eliminar filas con valores anómalos en la columna "id"
for val in not_num(movies.id.values):
    movies.drop(movies[(movies['id'] == val)].index, inplace=True)

not_num(movies.id.values)

[]

#### 2.3.3 Cambiar tipo de datos de columna `id`

In [17]:
# verificar y remplazar tipo de dato de columna "id"
id_dto = movies.dtypes['id']
print(f'El tipo original de datos de la columna "id": {id_dto}')

# Cambiar el tipo de datos de la columna 'id' de "object" a "int64"
movies['id'] = movies['id'].astype('int64')
id_dtn = movies.dtypes['id']
print(f'El nuevo tipo de datos de la columna "id": {id_dtn}')

El tipo original de datos de la columna "id": object
El nuevo tipo de datos de la columna "id": int64


### 2.4 Verificar valores no numéricos en columna `budget`de df movies

In [18]:
# Verificar valores no numéricos en columna "budget"
not_num(movies.budget.values)

[]

#### 2.4.1 Cambiar tipo de datos de columna `budget`

In [19]:
# verificar y remplazar tipo de dato de columna budget
budget_dto = movies.dtypes['budget']
print(f'El tipo original de datos de la columna "budget": {budget_dto}')

# Cambiar el tipo de datos de la columna 'budget' de "object" a "float64"
movies['budget'] = movies['budget'].astype('float64')
budget_dtn = movies.dtypes['budget']
print(f'El nuevo tipo de datos de la columna "budget": {budget_dtn}')

El tipo original de datos de la columna "budget": object
El nuevo tipo de datos de la columna "budget": float64


### 2.5 Eliminar columnas según requisitos del problema

#### 2.5.1 Eliminar columnas `video`, `imdb_id`, `adult`, `original_title`, `poster_path`, `homepage`

In [20]:
# Eliminar columnas por requisito del problema
movies.drop(['video', 'imdb_id', 'adult', 'original_title', 'poster_path', 'homepage'], axis=1, inplace=True)

### 2.6 Verificar registros con `id` duplicados

#### 2.6.1 Verificar si existe registros con `id` duplicados y borrarlos df movies

In [21]:
# Verificar si existe registros con "id" duplicados
print(f"Existen {movies[movies.duplicated('id')].shape[0]} duplicados")

Existen 17 duplicados


#### 2.6.2 Eliminar registros con `id` duplicados de df movies

In [22]:
# Eliminar registros con "id" duplicado df movies
movies = movies.drop_duplicates(movies.columns[movies.columns.isin(['id'])],
                        keep='first')

#### 2.6.3 Verificar si existe registros con `id` duplicados y borrarlos df credits

In [23]:
# Verificar si existe registros con "id" duplicados
print(f"Existen {credits[credits.duplicated('id')].shape[0]} duplicados")

Existen 7 duplicados


#### 2.6.4 Verificar si existe registros con `id` duplicados y borrarlos df credits

In [24]:
# Eliminar registros con "id" duplicado df credits
credits = credits.drop_duplicates(credits.columns[credits.columns.isin(['id'])],
                        keep='first')

#### Verificar shape de los df "movies" y "credits", transformar  y unir

In [25]:
# ids_faltantes = [x for x in set(movies['id']) if x not in set(credits['id'])]
# ids_faltantes

# movies.drop(movies[(movies['id'] == 401840)].index, inplace=True)  # ids_faltantes[0]

### 2.7 Realizar Merge de dfs `movies` y `credits` con `how = "left"` y `on = "id"`

In [26]:
# Verificar forma de los dataframes
movies.shape, credits.shape

((45433, 18), (45432, 3))

In [27]:
df = pd.merge(movies, credits, how='left', on='id')

In [28]:
df.shape

(45433, 20)

### 2.8 Aplanar Columnas anidadas

#### 2.8.1 Aplanar columnas `belongs_to_collection`, `genres`, `production_companies`, `production_countries` y `spoken_languages`

In [29]:
columns_to_flatten = [(df, 'belongs_to_collection') , (df, 'genres'), (df, 'production_companies'), (df, 'production_countries'), (df, 'spoken_languages')]

In [30]:
for tup in columns_to_flatten:
    flatten_column(tup[0], tup[1])

#### 2.8.2 Aplanar columna `crew`

In [31]:
# Convertir todos los valores faltantes en cadenas de listas vacías
df['crew'] = df['crew'].fillna('[]')
df['crew'] = df['crew'].apply(literal_eval)
df.iloc[0]['crew'][0]

{'credit_id': '52fe4284c3a36847f8024f49',
 'department': 'Directing',
 'gender': 2,
 'id': 7879,
 'job': 'Director',
 'name': 'John Lasseter',
 'profile_path': '/7EdqiNbr4FRjIhKHyPPdFfEEEFG.jpg'}

##### 2.8.2.1 Crear columna `director` con los nombres de las personas que ocupan ese trabajo en la columna crew

In [32]:
df['director'] = df['crew'].apply(get_director)

In [33]:
df['director'].head()

0      John Lasseter
1       Joe Johnston
2      Howard Deutch
3    Forest Whitaker
4      Charles Shyer
Name: director, dtype: object

#### 2.8.3 Aplanar columna `cast`

In [34]:
flatten_column(df, 'cast')

In [35]:
df['cast'][0]

['tom hanks',
 'tim allen',
 'don rickles',
 'jim varney',
 'wallace shawn',
 'john ratzenberger',
 'annie potts',
 'john morris',
 'erik von detten',
 'laurie metcalf',
 'r. lee ermey',
 'sarah freeman',
 'penn jillette']

#### 2.8.4 Eliminar columna `crew`

In [36]:
df.drop('crew', axis=1, inplace=True)

### 2.9 Imputar valores faltes columna `original_language`

In [37]:
df[['title', 'original_language']]

title original_language
0                        Toy Story                en
1                          Jumanji                en
2                 Grumpier Old Men                en
3                Waiting to Exhale                en
4      Father of the Bride Part II                en
...                            ...               ...
45428                       Subdue                fa
45429          Century of Birthing                tl
45430                     Betrayal                en
45431             Satan Triumphant                en
45432                     Queerama                en

[45433 rows x 2 columns]

#### 2.9.1 Imputar los valores faltantes del campo `original_language` con el valor iso639 del primer lenguaje encontrado en la columna `spoken_languages`

In [38]:
# crear columna con el cod iso-639 primer lenguaje encontrado en la lista de la columna "spoken_languages" (usando función inc_iso639)
df['first_lang'] = df['spoken_languages'].apply(lambda x: inc_iso639(x[0]) if len(x) > 0 else '')

In [39]:
# Verificar valores faltantes en columna "original_language"
df[df['original_language'].isnull()]

belongs_to_collection  budget                         genres      id  \
19566         no collection     0.0                  [documentary]  283101   
21589         no collection     0.0               [romance, drama]  103902   
22817         no collection     0.0                 [war, western]  359195   
32118         no collection     0.0                       [comedy]  147050   
37380         no collection     0.0                    [animation]  257095   
41017         no collection     0.0                  [documentary]  332742   
41842         no collection     0.0                  [documentary]  144410   
44027         no collection     0.0                        [drama]  380438   
44380         no collection     0.0       [animation, documentary]  381096   
44546         no collection     0.0                             []  381525   
44625         no collection     0.0  [action, war, drama, history]  104473   

      original_language                                           overview  \
19566               NaN  Documentary about the production of The Third ...   
21589               NaN  An Outback farmer takes in an Afghani woman wh...   
22817               NaN  A group of Union Army soldiers is charged with...   
32118               NaN  George and Gracie enter an elegant drawing roo...   
37380               NaN  The first fairy tale transformed into a full-l...   
41017               NaN  Until the late 1970s, the Pakistani city of La...   
41842               NaN  Two dancers perform in succession facing a sta...   
44027               NaN                                                NaN   
44380               NaN  The traditional crafts of crochet and knitting...   
44546               NaN  American investment fund buys Austrian compani...   
44625               NaN  Three military men, seen inside a fortificatio...   

      popularity                               production_companies  \
19566   0.017007  [studiocanal, british broadcasting corporation...   
21589   0.359818                  [new holland pictures, idtv film]   
22817   0.070647                       [associated producers (api)]   
32118   0.073418          [warner bros., the vitaphone corporation]   
37380   0.036841                       [ceskoslovenský státní film]   
41017   0.373688                                                 []   
41842    1.58434                     [edison manufacturing company]   
44027   0.001283                                                 []   
44380   0.067624                                                 []   
44546   0.030766                                                 []   
44625   0.471325                                        [star-film]   

                                    production_countries release_date  ...  \
19566  [austria, france, japan, united kingdom, unite...   2004-10-11  ...   
21589                                        [australia]   2007-08-04  ...   
22817                         [united states of america]   1960-03-31  ...   
32118                         [united states of america]   1929-10-31  ...   
37380                                   [czech republic]   1950-01-26  ...   
41017                                                 []   2015-11-13  ...   
41842                         [united states of america]   1895-04-01  ...   
44027                                           [france]   2001-01-01  ...   
44380                                                 []   2016-03-12  ...   
44546                                          [austria]   2016-01-27  ...   
44625                                           [france]   1897-01-01  ...   

       runtime              spoken_languages    status tagline  \
19566     95.0            [deutsch, english]  Released     NaN   
21589     94.0                     [english]  Released     NaN   
22817     69.0                     [english]  Released     NaN   
32118      8.0                     [english]  Released     NaN   
37380     87

In [40]:
# Imputar valores faltantes con valor de columna 'first_lang'
df['original_language'].fillna(df['first_lang'], inplace=True)

In [41]:
df.head()

belongs_to_collection      budget                        genres  \
0            toy story collection  30000000.0   [animation, comedy, family]   
1                   no collection  65000000.0  [adventure, fantasy, family]   
2       grumpy old men collection         0.0             [romance, comedy]   
3                   no collection  16000000.0      [comedy, drama, romance]   
4  father of the bride collection         0.0                      [comedy]   

      id original_language                                           overview  \
0    862                en  Led by Woody, Andy's toys live happily in his ...   
1   8844                en  When siblings Judy and Peter discover an encha...   
2  15602                en  A family wedding reignites the ancient feud be...   
3  31357                en  Cheated on, mistreated and stepped on, the wom...   
4  11862                en  Just when George Banks has recovered from his ...   

  popularity                               production_companies  \
0  21.946943                          [pixar animation studios]   
1  17.015539  [tristar pictures, teitler film, interscope co...   
2    11.7129                     [warner bros., lancaster gate]   
3   3.859495           [twentieth century fox film corporation]   
4   8.387519       [sandollar productions, touchstone pictures]   

         production_countries release_date  ...  runtime     spoken_languages  \
0  [united states of america]   1995-10-30  ...     81.0            [english]   
1  [united states of america]   1995-12-15  ...    104.0  [english, français]   
2  [united states of america]   1995-12-22  ...    101.0            [english]   
3  [united states of america]   1995-12-22  ...    127.0            [english]   
4  [united states of america]   1995-02-10  ...    106.0            [english]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title vote_average  vote_count  \
0                    Toy Story          7.7      5415.0   
1                      Jumanji          6.9      2413.0   
2             Grumpier Old Men          6.5        92.0   
3            Waiting to Exhale          6.1        34.0   
4  Father of the Bride Part II          5.7       173.0   

                                                cast         director  \
0  [tom hanks, tim allen, don rickles, jim varney...    John Lasseter   
1  [robin williams, jonathan hyde, kirsten dunst,...     Joe Johnston   
2  [walter matthau, jack lemmon, ann-margret, sop...    Howard Deutch   
3  [whitney houston, angela bassett, loretta devi...  Forest Whitaker   
4  [steve martin, diane keaton, martin short, kim...    Charles Shyer   

  first_lang  
0         en  
1         en  
2         en  
3         en  
4         en  

[5 rows x 21 columns]

In [42]:
df[df['title'] == 'Song of Lahore']

belongs_to_collection  budget         genres      id original_language  \
41017         no collection     0.0  [documentary]  332742                ur   

                                                overview popularity  \
41017  Until the late 1970s, the Pakistani city of La...   0.373688   

      production_companies production_countries release_date  ...  runtime  \
41017                   []                   []   2015-11-13  ...     82.0   

              spoken_languages    status tagline           title vote_average  \
41017  [اردو, english, ਪੰਜਾਬੀ]  Released     NaN  Song of Lahore          6.5   

       vote_count  cast       director first_lang  
41017         2.0    []  Andy Schocken         ur  

[1 rows x 21 columns]

In [43]:
df.isnull().sum()*100/df.shape[0]

belongs_to_collection     0.000000
budget                    0.000000
genres                    0.000000
id                        0.000000
original_language         0.000000
overview                  2.099795
popularity                0.006603
production_companies      0.000000
production_countries      0.000000
release_date              0.191491
revenue                   0.006603
runtime                   0.572271
spoken_languages          0.000000
status                    0.184888
tagline                  55.096516
title                     0.006603
vote_average              0.006603
vote_count                0.006603
cast                      0.000000
director                  1.954526
first_lang                0.609689
dtype: float64

### Eliminar otras columnas no requeridas para el análisis

In [44]:
df.drop(['first_lang', 'status', 'popularity', 'spoken_languages', 'vote_average', 'vote_count'], axis=1, inplace=True)
df.head()

belongs_to_collection      budget                        genres  \
0            toy story collection  30000000.0   [animation, comedy, family]   
1                   no collection  65000000.0  [adventure, fantasy, family]   
2       grumpy old men collection         0.0             [romance, comedy]   
3                   no collection  16000000.0      [comedy, drama, romance]   
4  father of the bride collection         0.0                      [comedy]   

      id original_language                                           overview  \
0    862                en  Led by Woody, Andy's toys live happily in his ...   
1   8844                en  When siblings Judy and Peter discover an encha...   
2  15602                en  A family wedding reignites the ancient feud be...   
3  31357                en  Cheated on, mistreated and stepped on, the wom...   
4  11862                en  Just when George Banks has recovered from his ...   

                                production_companies  \
0                          [pixar animation studios]   
1  [tristar pictures, teitler film, interscope co...   
2                     [warner bros., lancaster gate]   
3           [twentieth century fox film corporation]   
4       [sandollar productions, touchstone pictures]   

         production_countries release_date      revenue  runtime  \
0  [united states of america]   1995-10-30  373554033.0     81.0   
1  [united states of america]   1995-12-15  262797249.0    104.0   
2  [united states of america]   1995-12-22          0.0    101.0   
3  [united states of america]   1995-12-22   81452156.0    127.0   
4  [united states of america]   1995-02-10   76578911.0    106.0   

                                             tagline  \
0                                                NaN   
1          Roll the dice and unleash the excitement!   
2  Still Yelling. Still Fighting. Still Ready for...   
3  Friends are the people who let you be yourself...   
4  Just When His World Is Back To Normal... He's ...   

                         title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                                cast         director  
0  [tom hanks, tim allen, don rickles, jim varney...    John Lasseter  
1  [robin williams, jonathan hyde, kirsten dunst,...     Joe Johnston  
2  [walter matthau, jack lemmon, ann-margret, sop...    Howard Deutch  
3  [whitney houston, angela bassett, loretta devi...  Forest Whitaker  
4  [steve martin, diane keaton, martin short, kim...    Charles Shyer

In [45]:
df.isnull().sum()*100/df.shape[0]

belongs_to_collection     0.000000
budget                    0.000000
genres                    0.000000
id                        0.000000
original_language         0.000000
overview                  2.099795
production_companies      0.000000
production_countries      0.000000
release_date              0.191491
revenue                   0.006603
runtime                   0.572271
tagline                  55.096516
title                     0.006603
cast                      0.000000
director                  1.954526
dtype: float64

### 2.10 Crear la columna con el retorno de inversión

#### 2.10.1 imputar con **0** valores nulos de columna `revenue`

In [46]:
df['revenue'].fillna(0, inplace=True)

#### 2.10.2 Crear la columna con el retorno de inversión, llamada return con los campos revenue y budget, dividiendo estas dos últimas revenue / budget, cuando no hay datos disponibles para calcularlo, deberá tomar el valor 0

In [47]:
df['return'] = df.apply(lambda x: x['revenue'] / x['budget'] if x['budget'] != 0 else 0, axis=1)

In [48]:
df.isnull().sum()*100/df.shape[0]

belongs_to_collection     0.000000
budget                    0.000000
genres                    0.000000
id                        0.000000
original_language         0.000000
overview                  2.099795
production_companies      0.000000
production_countries      0.000000
release_date              0.191491
revenue                   0.000000
runtime                   0.572271
tagline                  55.096516
title                     0.006603
cast                      0.000000
director                  1.954526
return                    0.000000
dtype: float64

#### 2.10.3 Crear columna `year` 

In [49]:
# Verificar información de columna "release_date"
df['release_date'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 45433 entries, 0 to 45432
Series name: release_date
Non-Null Count  Dtype 
--------------  ----- 
45346 non-null  object
dtypes: object(1)
memory usage: 355.1+ KB


In [50]:
# Cambiar a tipo datetime columna "release_date" y se crea la columna release_year 
df['release_date'] = pd.to_datetime(df['release_date'], format= '%Y-%m-%d', errors='coerce')
df['release_year'] = df['release_date'].dt.year

In [51]:
# Pasar a int los datos de la columna release_year ya que la transformación los entrego como float64
df['release_year'] = df['release_year'].astype('Int64')
df[['release_date', 'release_year']].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45433 entries, 0 to 45432
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   release_date  45346 non-null  datetime64[ns]
 1   release_year  45346 non-null  Int64         
dtypes: Int64(1), datetime64[ns](1)
memory usage: 754.4 KB


In [52]:
df[['release_date', 'release_year']].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45433 entries, 0 to 45432
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   release_date  45346 non-null  datetime64[ns]
 1   release_year  45346 non-null  Int64         
dtypes: Int64(1), datetime64[ns](1)
memory usage: 754.4 KB


### 2.11 Eliminar columna `tagline`

In [53]:
# Se elimina la columna "tagline" ya que presenta un 55.10% de datos nulos
df.drop(['tagline'], axis=1, inplace=True)

### 2.12 Eliminar filas con datos nulos

In [54]:
# Se eliminan las filas que quedan y contienen datos nulos
df_clean = df.dropna()

In [55]:
print('Cantidad de registros dataset original sin filas repetidas: ', df.shape[0])
print('Cantidad de registros dataset limpio: ', df_clean.shape[0])
print('*' * 80)
print(f'El DataFrame Limpio (df_clean) mantiene el {round(df_clean.shape[0]*100/df.shape[0], 2)}% de los datos originales')
print('*' * 80)

Cantidad de registros dataset original sin filas repetidas:  45433
Cantidad de registros dataset limpio:  43641
********************************************************************************
El DataFrame Limpio (df_clean) mantiene el 96.06% de los datos originales
********************************************************************************


In [56]:
df_clean.isnull().sum()*100/df_clean.shape[0]

belongs_to_collection    0.0
budget                   0.0
genres                   0.0
id                       0.0
original_language        0.0
overview                 0.0
production_companies     0.0
production_countries     0.0
release_date             0.0
revenue                  0.0
runtime                  0.0
title                    0.0
cast                     0.0
director                 0.0
return                   0.0
release_year             0.0
dtype: float64

## 3. Generar archivos para trabajar limpios

In [58]:
# Generar archivo movies_cleaned.csv
# df_clean.to_csv('datasets/movies_cleaned.csv', index=False, encoding='utf-8')

# Generar archivo movies_reduced.csv
# df.to_csv('datasets/movies_reduced.csv', index=False, encoding='utf-8')

## <span style="color:green">Final ETL</span>

In [59]:
import pandas as pd
df = pd.read_csv('datasets/movies_reduced.csv')

In [60]:
def get_lang(idioma:str):
    '''
    Función para obtener el codigo iso639 de un pais o el nombre del mismo escrito en el idioma 
    '''
    lang = {
            'aa' : 'afar',
            'ab' : 'abjasio',  # (o abjasiano)
            'ae' : 'avéstico',
            'af' : 'afrikáans',
            'ak' : 'akano',
            'am' : 'amhárico',
            'an' : 'aragonés',
            'ar' : 'العربية',
            'as' : 'asamés',
            'av' : 'avar',  # (o ávaro)
            'ay' : 'aimara',
            'az' : 'azerí',
            'ba' : 'baskir',
            'be' : 'беларуская мова',
            'bg' : 'български език',
            'bh' : 'bhoyapurí',
            'bi' : 'bislama',
            'bm' : 'bamanankan',
            'bn' : 'বাংলা',
            'bo' : 'tibetano',
            'br' : 'bretón',
            'bs' : 'bosanski',
            'ca' : 'català',
            'ce' : 'checheno',
            'ch' : 'chamorro',
            'co' : 'corso',
            'cr' : 'cree',
            'cs' : 'český',
            'cu' : 'eslavo',  # eclesiástico antiguo
            'cv' : 'chuvasio',
            'cy' : 'cymraeg',
            'da' : 'dansk',
            'de' : 'deutsch',
            'dv' : 'maldivo',  # (o dhivehi)
            'dz' : 'dzongkha',
            'ee' : 'ewé',
            'el' : 'ελληνικά',  # (moderno)
            'en' : 'english',
            'eo' : 'esperanto',
            'es' : 'español',  # (o castellano)
            'et' : 'eesti',
            'eu' : 'euskera',
            'fa' : 'فارسی',
            'ff' : 'fula',
            'fi' : 'suomi',  # (o finlandés)
            'fj' : 'fiyiano',  # (o fiyi)
            'fo' : 'feroés',
            'fr' : 'français',
            'fy' : 'frisón',  # (o frisio)
            'ga' : 'gaeilge',  # (o gaélico)
            'gd' : 'gaélico',  # escocés
            'gl' : 'galego',
            'gn' : 'guaraní',
            'gu' : 'guyaratí',  # (o gujaratí)
            'gv' : 'manés',  # (gaélico manés o de Isla de Man)
            'ha' : 'hausa',
            'he' : 'עִבְרִית',
            'hi' : 'हिन्दी',  # (o hindú)
            'ho' : 'hiri',  # motu
            'hr' : 'hrvatski',
            'ht' : 'haitiano',
            'hu' : 'magyar',
            'hy' : 'armenio',
            'hz' : 'herero',
            'ia' : 'interlingua',
            'id' : 'bahasa indonesia',
            'ie' : 'occidental',
            'ig' : 'igbo',
            'ii' : 'yi de Sichuán',
            'ik' : 'iñupiaq',
            'io' : 'ido',
            'is' : 'íslenska',
            'it' : 'italiano',
            'iu' : 'inuktitut',  # (o inuit)
            'ja' : '日本語',
            'jv' : 'javanés',
            'ka' : 'ქართული',
            'kg' : 'kongo',  # (o kikongo)
            'ki' : 'kikuyu',
            'kj' : 'kuanyama',
            'kk' : 'қазақ',  # (o kazajio)
            'kl' : 'groenlandés',  # (o kalaallisut)
            'km' : 'camboyano',  # o jemer)
            'kn' : 'Kannada',
            'ko' : '한국어/조선말',
            'kr' : 'kanuri',
            'ks' : 'cachemiro',  # (o cachemir)
            'ku' : 'kurdo',
            'kv' : 'komi',
            'kw' : 'córnico',
            'ky' : 'kirghiz',
            'la' : 'latín',
            'lb' : 'luxemburgués',
            'lg' : 'luganda',
            'li' : 'limburgués',
            'ln' : 'lingala',
            'lo' : 'lao',
            'lt' : 'lietuvi\x9akai',
            'lu' : 'luba-katanga',  # (o chiluba)
            'lv' : 'latviešu',
            'mg' : 'malgache',  # (o malagasy)
            'mh' : 'marshalés',
            'mi' : 'maorí',
            'mk' : 'macedonio',
            'ml' : 'malayalam',
            'mn' : 'mongol',
            'mr' : 'maratí',
            'ms' : 'bahasa melayu',
            'mt' : 'malti',
            'my' : 'birmano',
            'na' : 'nauruano',
            'nb' : 'bokmål',  # bokmål
            'nd' : 'ndebele del norte',
            'ne' : 'nepalí',
            'ng' : 'ndonga',
            'nl' : 'nederlands',  # (u holandés)
            'nn' : 'nynorsk',
            'no' : 'norsk',
            'nr' : 'ndebele del sur',
            'nv' : 'navajo',
            'ny' : 'chichewa',
            'oc' : 'occitano',
            'oj' : 'ojibwa',
            'om' : 'oromo',
            'or' : 'oriya',
            'os' : 'osético',  # (u osetio, u oseta)
            'pa' : 'ਪੰਜਾਬੀ', # (o penyabi)
            'pi' : 'pali',
            'pl' : 'polski',
            'ps' : 'پښتو',  # (o pastún, o pashto)
            'pt' : 'português',
            'qc' : 'quechua',
            'rm' : 'romanche',
            'rn' : 'kirundi',
            'ro' : 'română',
            'ru' : 'pусский',
            'rw' : 'kinyarwanda',  # (o kiñaruanda)
            'sa' : 'sánscrito',
            'sc' : 'sardo',
            'sd' : 'sindhi',
            'se' : 'sami septentrional',
            'sg' : 'sango',
            'si' : 'cingalés',
            'sk' : 'slovenčina',
            'sl' : 'slovenščina',
            'sm' : 'samoano',
            'sn' : 'shona',
            'so' : 'somalí',
            'sq' : 'shqip',
            'sr' : 'srpski',
            'ss' : 'suazi',  # (o swati, o siSwati)
            'st' : 'sesotho',
            'su' : 'sundanés',  # (o sondanés)
            'sv' : 'svenska',
            'sw' : 'kiswahili',
            'ta' : 'தமிழ்',
            'te' : 'తెలుగు',
            'tg' : 'tayiko',
            'th' : 'ภาษาไทย',
            'ti' : 'tigriña',
            'tk' : 'turcomano',
            'tl' : 'tagalo',
            'tn' : 'setsuana',
            'to' : 'tongano',
            'tr' : 'türkçe',
            'ts' : 'tsonga',
            'tt' : 'tártaro',
            'tw' : 'twi',
            'ty' : 'tahitiano',
            'ug' : 'uigur',
            'uk' : 'український',
            'ur' : 'اردو',
            'uz' : 'ozbek',
            've' : 'venda',
            'vi' : 'tiếng việt',
            'vo' : 'volapük',
            'wa' : 'valón',
            'wo' : 'wolof',
            'xh' : 'xhosa',
            'yi' : 'yídish',  # (o yidis, o yiddish)
            'yo' : 'yoruba',
            'za' : 'chuan',  # (o chuang, o zhuang)
            'zh' : '普通话',
            'zu' : 'isizulu',
            'cn' : '广州话 / 廣州話',
            'Sin Idioma': 'no language'
            }
    
    if len(idioma)>2:
        for k, v in lang.items():
            if v == idioma.lower():
                return k
    else:
        for k, v in lang.items():

            if k == idioma.lower():
                return v

In [61]:
idioma = 'zh'

def peliculas_idioma(lan):
    idioma = get_lang(lan)
    if len(idioma) > 2:
        idioma, lan = lan, idioma
        
    resul = df.loc[df['original_language'] == idioma].shape[0]# df_clean['original_language'][df_clean['original_language'] == idioma].count()
    return{
        'idioma':f'{idioma}-{lan}',
        'cantidad': resul
    }
    
peliculas_idioma(idioma)

{'idioma': 'zh-普通话', 'cantidad': 409}

In [62]:
pelicula = 'Rocky II'
def peliculas_duracion(pelicula):
    try:        
        resul = df.loc[df['title'] == pelicula]
        movie_length = resul['runtime'].values[0]
        year = resul['release_year'].values[0]
        return movie_length, year
    except:
        return 'Sin resultado'

peliculas_duracion(pelicula)

'Sin resultado'

In [63]:
franq = 'toy story collection'
def franquicia(franquicia):
        
        '''
         Retornar la cantidad de peliculas, ganancia total y ganancia promedio de la franquicia pasada como parámetro.

         Por criterio del problema se tomará como ganancia la columna "revenue" 

         Parametros
         ----------

         franquicia: str el nombre de la franquisia
        '''
        
        resul = df.loc[df['belongs_to_collection'] == franquicia]
        
        if len(resul) == 0:
            return False
        else:
            amount_movies = resul.shape[0]
            total_revenue = round(resul.revenue.sum()/1000000, 2)
            mean_revenue = round(resul.revenue.mean()/1000000, 2)
            return amount_movies, total_revenue, mean_revenue
        
print(franquicia(franq))

(3, 1937.89, 645.96)


In [64]:
p = 'philippines'

def pelicula_pais(pais: str):
                          
    '''
    Retornar la cantidad de peliculas producidas en el país pasado como parámetro.

    Parametros
    ----------

    pais: str el nombre del país según se encuentra en la base de datos
    '''

    df_coun = df.loc[:,['title','production_countries']]
    df_coun['production_countries'] = df_coun['production_countries'].apply(eval)
    df_coun = df_coun.explode('production_countries')
    resul = df_coun[df_coun['production_countries'] == pais].shape[0]
    
    return resul
    
    
pelicula_pais(p)

83

In [65]:
df_coun = df.loc[:,['title','production_countries']]
df_coun['production_countries'] = df_coun['production_countries'].apply(eval)
df_coun = df_coun.explode('production_countries')
resul = df_coun[df_coun['production_countries'] == 'venezuela'].shape[0]
resul

19

In [147]:
p = 'tristar pictures'
def productoras_exitosas(productora):
    
    '''
    Retorna el revunue total y la cantidad de peliculas que realizo la productora pasada por parámetro 


    Parametros
    ----------

    productora: str el nombre de la productora según se encuentra en la base de datos
    '''

    df_produc = df.loc[:,['production_companies','revenue']]
    df_produc['production_companies'] = df_produc['production_companies'].apply(eval)
    df_produc = df_produc.explode('production_companies')

    total_revenue = df_produc[df_produc['production_companies'] == productora]['revenue'].sum()
    amount_movies = df_produc[df_produc['production_companies'] == productora].shape[0]

    return total_revenue, amount_movies

productoras_exitosas(p)        

(8022366764.0, 197)

In [72]:
df_produc = df.loc[:,['production_companies','revenue']]
df_produc['production_companies'] = df_produc['production_companies'].apply(eval)
df_produc = df_produc.explode('production_companies')

total_revenue = df_produc[df_produc['production_companies'] == 'tristar pictures']['revenue'].sum()
amount_movies = df_produc[df_produc['production_companies'] == 'tristar pictures'].shape[0]
total_revenue, amount_movies

(8022366764.0, 197)

In [69]:
df_produc

production_companies      revenue
0        pixar animation studios  373554033.0
1               tristar pictures  262797249.0
1                   teitler film  262797249.0
1      interscope communications  262797249.0
2                   warner bros.          0.0
...                          ...          ...
45428                        NaN          0.0
45429                sine olivia          0.0
45430    american world pictures          0.0
45431                  yermoliev          0.0
45432                        NaN          0.0

[82335 rows x 2 columns]

In [85]:
list(df_produc['production_companies'].unique())

['pixar animation studios',
 'tristar pictures',
 'teitler film',
 'interscope communications',
 'warner bros.',
 'lancaster gate',
 'twentieth century fox film corporation',
 'sandollar productions',
 'touchstone pictures',
 'regency enterprises',
 'forward pass',
 'paramount pictures',
 'scott rudin productions',
 'mirage enterprises',
 'constellation entertainment',
 'worldwide',
 'mont blanc entertainment gmbh',
 'walt disney pictures',
 'universal pictures',
 'imperial entertainment',
 'signature entertainment',
 'united artists',
 'eon productions',
 'columbia pictures',
 'castle rock entertainment',
 'enigma pictures',
 'amblin entertainment',
 'amblimation',
 'hollywood pictures',
 'cinergi pictures entertainment',
 'le studio canal+',
 'laurence mark productions',
 'metro-goldwyn-mayer (mgm)',
 'carolco pictures',
 'légende entreprises',
 'syalis da',
 'de fina-cappa',
 'columbia pictures corporation',
 'miramax films',
 'a band apart',
 'o entertainment',
 'morgan creek produ

In [95]:
df_produc[df_produc['production_companies'] == 'irs media']

production_companies    revenue
462              irs media        0.0
3421             irs media        0.0
3605             irs media  1543112.0
7521             irs media        0.0
9853             irs media        0.0
16349            irs media        0.0
22202            irs media        0.0
39072            irs media        0.0

In [ ]:
director = ''
def director(nombre_director):
    '''
    retorna el éxito medido a través del retorno del director pasado por parámetro. 

    La función retornará ademas del exito del director lo siguiente en una lista:
    nombre de cada película, fecha de lanzamiento, retorno individual, costo y ganancia de la misma. 

    '''
    df_director = df.loc[:,['director', 'return', 'budget','revenue']]
    budget_director = df_director[df_director['director'] == nombre_director]['budget'].sum()
    revenue_director = df_director[df_director['director'] == nombre_director]['revenue'].sum()
    return_direct = revenue_director/budget_director
    

In [144]:
df_director = df.loc[:,['director', 'title', 'release_date','return', 'budget','revenue']]
budget_director = df_director[df_director['director'] == nombre_director]['budget'].sum()
revenue_director = df_director[df_director['director'] == nombre_director]['revenue'].sum()
return_direct = revenue_director/budget_director

In [145]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45433 entries, 0 to 45432
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   belongs_to_collection  45433 non-null  object 
 1   budget                 45433 non-null  float64
 2   genres                 45433 non-null  object 
 3   id                     45433 non-null  int64  
 4   original_language      45433 non-null  object 
 5   overview               44479 non-null  object 
 6   production_companies   45433 non-null  object 
 7   production_countries   45433 non-null  object 
 8   release_date           45346 non-null  object 
 9   revenue                45433 non-null  float64
 10  runtime                45173 non-null  float64
 11  title                  45430 non-null  object 
 12  cast                   45433 non-null  object 
 13  director               44545 non-null  object 
 14  return                 45433 non-null  float64
dtypes:

In [155]:
nombre_director ='Ang Lee'
df_director[df_director['director'] == nombre_director]

director                            title release_date     return  \
16     Ang Lee            Sense and Sensibility   1995-12-13   8.181818   
228    Ang Lee              Eat Drink Man Woman   1994-08-03   0.000000   
294    Ang Lee                    Pushing Hands   1992-02-01   0.380805   
1558   Ang Lee                    The Ice Storm   1997-09-27   0.446559   
2999   Ang Lee              Ride with the Devil   1999-11-24   0.016713   
3870   Ang Lee   Crouching Tiger, Hidden Dragon   2000-10-01  12.560337   
6384   Ang Lee                             Hulk   2003-06-19   1.790952   
6870   Ang Lee              The Wedding Banquet   1993-08-04   6.933459   
10447  Ang Lee               Brokeback Mountain   2005-09-02  12.717412   
12081  Ang Lee                    Lust, Caution   2007-09-28   0.000000   
14099  Ang Lee                 Taking Woodstock   2009-09-03   0.000000   
19726  Ang Lee                       Life of Pi   2012-11-20   5.075138   
25914  Ang Lee                 The Hire: Chosen   2001-01-01   0.000000   
40465  Ang Lee  Billy Lynn's Long Halftime Walk   2016-11-10   0.773275   

            budget      revenue  
16      16500000.0  135000000.0  
228            0.0    7294403.0  
294       400000.0     152322.0  
1558    18000000.0    8038061.0  
2999    38000000.0     635096.0  
3870    17000000.0  213525736.0  
6384   137000000.0  245360480.0  
6870     1000000.0    6933459.0  
10447   14000000.0  178043761.0  
12081   15000000.0          0.0  
14099          0.0          0.0  
19726  120000000.0  609016565.0  
25914          0.0          0.0  
40465   40000000.0   30930984.0

In [156]:
print(len(df_director[df_director['director'] == nombre_director]))

14


In [157]:
budget_director = df_director[df_director['director'] == nombre_director]['budget'].sum()
budget_director

416900000.0

In [158]:
revenue_director = df_director[df_director['director'] == nombre_director]['revenue'].sum()
revenue_director

1434930867.0

In [159]:
return_direct = revenue_director/budget_director
return_direct

3.441906613096666

In [160]:
movies_director = df_director[df_director['director'] == nombre_director].values
records_director = []

for l in range(len(movies_director)):
    dict_director = {
        'title': movies_director[l][1],
        'release_date': movies_director[l][2],
        'return': movies_director[l][3],
        'budget': movies_director[l][4],
        'revenue': movies_director[l][5]    
        }
    records_director.append(dict_director)
records_director

[{'title': 'Sense and Sensibility',
  'release_date': '1995-12-13',
  'return': 8.181818181818182,
  'budget': 16500000.0,
  'revenue': 135000000.0},
 {'title': 'Eat Drink Man Woman',
  'release_date': '1994-08-03',
  'return': 0.0,
  'budget': 0.0,
  'revenue': 7294403.0},
 {'title': 'Pushing Hands',
  'release_date': '1992-02-01',
  'return': 0.380805,
  'budget': 400000.0,
  'revenue': 152322.0},
 {'title': 'The Ice Storm',
  'release_date': '1997-09-27',
  'return': 0.4465589444444444,
  'budget': 18000000.0,
  'revenue': 8038061.0},
 {'title': 'Ride with the Devil',
  'release_date': '1999-11-24',
  'return': 0.0167130526315789,
  'budget': 38000000.0,
  'revenue': 635096.0},
 {'title': 'Crouching Tiger, Hidden Dragon',
  'release_date': '2000-10-01',
  'return': 12.560337411764706,
  'budget': 17000000.0,
  'revenue': 213525736.0},
 {'title': 'Hulk',
  'release_date': '2003-06-19',
  'return': 1.790952408759124,
  'budget': 137000000.0,
  'revenue': 245360480.0},
 {'title': 'The 

In [125]:
list(df_director['director'].unique())

['John Lasseter',
 'Joe Johnston',
 'Howard Deutch',
 'Forest Whitaker',
 'Charles Shyer',
 'Michael Mann',
 'Sydney Pollack',
 'Peter Hewitt',
 'Peter Hyams',
 'Martin Campbell',
 'Rob Reiner',
 'Mel Brooks',
 'Simon Wells',
 'Oliver Stone',
 'Renny Harlin',
 'Martin Scorsese',
 'Ang Lee',
 'Allison Anders',
 'Steve Oedekerk',
 'Joseph Ruben',
 'Barry Sonnenfeld',
 'Jon Amiel',
 'Richard Donner',
 'Victor Salva',
 'Mike Figgis',
 'Oliver Parker',
 'Lesli Linka Glatter',
 'Roger Michell',
 'Jean-Pierre Jeunet',
 'Zhang Yimou',
 'John N. Smith',
 'Terry Gilliam',
 'Jean-Jacques Annaud',
 'Chris Noonan',
 'Christopher Hampton',
 'Tim Robbins',
 'Stephen Low',
 'Andy Tennant',
 'Amy Heckerling',
 'Darrell James Roodt',
 'Richard Loncraine',
 'Albert Hughes',
 'Michael Hoffman',
 'Paul W.S. Anderson',
 'Gus Van Sant',
 'Jocelyn Moorhouse',
 'David Fincher',
 'Mike Gabriel',
 'Patricia Rozema',
 'Bryan Singer',
 'Richard W. Munchkin',
 'Woody Allen',
 'Gianni Amelio',
 'Holly Goldberg Sloan

In [ ]:
{
    'director': 'Jean-Jacques Annaud',
    'return': 4.02,
    'movies':[
        {
        'title': 'Dead Man Walking',
        'release_date': '1995-12-29',
        'return': 3.578512,
        'budget': 11000000.0,
        'revenue': 39363635.0    
        },
        {
        'title': 'Dead Man Walking',
        'release_date': '1992-09-04',
        'return': 0.00,
        'budget': 0.00,
        'revenue': 0.00    
        },
        {
        'title': 'Dead Man Walking',
        'release_date': '1999-12-10',
        'return': 0.00,
        'budget': 0.00,
        'revenue': 0.00    
        }
    ]

}